### DISCUSSION TITLE
External Machine Translation

### DISCUSSION THREAD

**Discussion by AgentOren** [tolgee/tolgee-platform#2324] [Q&A]
Dear Community,

In one of our project, we currently use ChatGPT with a fine-tuned model to translate texts in our specific domain.
So far we have very good results, much better than other machine translation services we tried, due to training/fine-tuning based on human-translated content in our domain.

I'm currently pushing new translations to Tolgee using the ["Create new key" API](https://tolgee.io/api/create-2). Since new texts are added frequently and reviewed by technical writers and translators, I would like to have the same effect as auto-translation in Tolgee: the texts are marked as "Machine Translated" with a special icon, to distinguish between machine-translated and human-translated texts.

I thought about two paths to achieve this:
1. Develop a custom "Machine Translation" service for Tolgee: Unfortunately, I could not find documents about this.
2. Push translated texts like I do today, but mark them as "machine translated". Unfortunately, I could not find an API that does this.

Has any of you used an external MT with Tolgee?
If so, how?

**Comment by JanCizmar** [comment]
Hi Oren, we were thinking about adding support for custom translation services, also adding the endpoint for marking translation "machine translated" would be pretty simple.  :) 

Anyway, it would be great to have a call with you, we always planned to add fine-tuning as a native feature and it would be great to know how far you've got. Would you be open to? This is [my Calendly](https://tolg.ee/jan) 🚀

**Reply by AgentOren** [reply]
Hi Jan!

Actually the usage of our fine-tuning model attempts are currently on-hold. The reason is that fine-tuning is currently only possible with older GPT-3.5 models. The good results we've seen are actually coming from stock GPT-4-Turbo and GPT-4o models, using a custom prompt (sorry for misleading). We're waiting for fine-tuning to become available for GPT-4 or newer models to resume our fine-tuning attempts.

Our custom prompts works better than other MT services we've tried, as the prompts include special instructions for handling our texts, e.g.: some texts contain markdown format that must be preserved, including code-like blocks that must never be translated or modified.

MT services also don't allow us to provide per-language or per-context instructions, such as translation of internal terms and industry Jargon (e.g. there are Hebrew translations for terms like "Internet", "Logs", "Ports" but they're rarely used by IT personnel, who prefer a less-formal phonetic copy of these English terms).

I'll be happy to share some knowledge with you, let me know if you still want to talk/chat.

**Reply by brianbento** [reply]
Looking for the same thing. Can't find any documentation. @JanCizmar any news? Anything changed? Can this be done via a plugin?

Really excited to use Tolgee if I can customize choices for translation.

**Reply by JanCizmar** [reply]
Hello! 

We are currently having on the top priority list. 

By 17th March we start working in 6 weeks development cycles approach and adding this should be implemented in the first cycle. I cannot tell you 100% it when it will be released, but my current ETA is end beginning of May. 

In this feature we plan to add 
- some AI playground which will help you to customize your prompt.
- enable you to add your own API keys to different models
- prepare for adding more LLM features like automated key description generation or rewording of the strings

**Reply by btalin** [reply]
@JanCizmar how far along are you with that? tried to add a together.ai (OpenAI based) key but I cant find any documentations there really and I would like to add there some LLMs.

**Reply by stepan662** [reply]
Hi @btalin, custom LLM providers and AI playground are released
https://docs.tolgee.io/platform/projects_and_organizations/llm-providers#self-hosted-server-configuration
https://docs.tolgee.io/platform/translation_process/ai-playground

**Reply by btalin** [reply]
Actually that is how far I got. And it doesn't even tell me where I should put all the stuff 😂 tried with ChatGPT to add it to .env but not working either (opened another discussion here as it's not really straight forward ... Simple GUI is explained with thousands of pictures but here it's in server configuration stuff etc. And just with a simple code snippet explained like it's so self explaining)
